From 3d4da77caca0a1916dc3b06b74bab91e6579dcbc Mon Sep 17 00:00:00 2001
From: Henri Yandell <bayard@apache.org>
Date: Thu, 5 Oct 2006 18:29:09 +0000
Subject: [PATCH] Fixing ArrayOutOfBoundsException bug reported in
 #VALIDATOR-202. This fix is pretty simple - 'Create an array of size # of .s
 plus one', but the implementation of that isn't very beautiful.

git-svn-id: https://svn.apache.org/repos/asf/jakarta/commons/proper/validator/trunk@453308 13f79535-47bb-0310-9956-ffa450edef68
---
 .../org/apache/commons/validator/UrlValidator.java    | 11 ++++++++++-
 src/test/org/apache/commons/validator/UrlTest.java    |  8 +++++++-
 2 files changed, 17 insertions(+), 2 deletions(-)

diff --git a/src/share/org/apache/commons/validator/UrlValidator.java b/src/share/org/apache/commons/validator/UrlValidator.java
index 86d1d8820..a0f34f6ed 100644
--- a/src/share/org/apache/commons/validator/UrlValidator.java
+++ b/src/share/org/apache/commons/validator/UrlValidator.java
@@ -355,7 +355,16 @@ protected boolean isValidAuthority(String authority) {
 
 //rightmost hostname will never start with a digit.
         if (hostname) {
-            String[] domainSegment = new String[10];
+            // LOW-TECH FIX FOR VALIDATOR-202
+            // TODO: Rewrite to use ArrayList and .add semantics: see VALIDATOR-203
+            char[] chars = hostIP.toCharArray();
+            int size = 1;
+            for(int i=0; i<chars.length; i++) {
+                if(chars[i] == '.') {
+                    size++;
+                }
+            }
+            String[] domainSegment = new String[size];
             boolean match = true;
             int segmentCount = 0;
             int segmentLength = 0;
diff --git a/src/test/org/apache/commons/validator/UrlTest.java b/src/test/org/apache/commons/validator/UrlTest.java
index c2113054e..ebd4e8b9b 100644
--- a/src/test/org/apache/commons/validator/UrlTest.java
+++ b/src/test/org/apache/commons/validator/UrlTest.java
@@ -135,6 +135,12 @@ public void testIsValid(Object[] testObjects, int options) {
       }
    }
 
+   public void testValidator202() {
+       String[] schemes = {"http","https"};
+       UrlValidator urlValidator = new UrlValidator(schemes, UrlValidator.NO_FRAGMENTS);
+       urlValidator.isValid("http://www.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.logoworks.comwww.log");
+   }
+
    static boolean incrementTestPartsIndex(int[] testPartsIndex, Object[] testParts) {
       boolean carry = true;  //add 1 to lowest order part.
       boolean maxIndex = true;
@@ -278,4 +284,4 @@ public static void main(String[] argv) {
                             new TestPair("telnet", false)};
 
 
-}
\ No newline at end of file
+}
